File: mmupnp/src/main/java/net/mm2d/upnp/ControlPointFactory.java
/*
 * Copyright (c) 2018 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import net.mm2d.upnp.internal.impl.ControlPointImpl;
import net.mm2d.upnp.internal.impl.DiFactory;

import java.net.NetworkInterface;
import java.util.Collection;
import java.util.Collections;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * ControlPointのインスタンスを作成するFactory。
 *
 * <p>将来的にControlPointもインターフェースに変更するため、
 * ControlPointのコンストラクタはDeprecatedとしている。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public final class ControlPointFactory {
    /**
     * ControlPointの初期化パラメータ
     */
    public static class Params {
        @Nonnull
        private Protocol mProtocol = Protocol.DEFAULT;
        @Nullable
        private Collection<NetworkInterface> mInterfaces;
        @Nullable
        private TaskExecutor mCallbackExecutor;

        private boolean mNotifySegmentCheckEnabled;

        /**
         * 使用するプロトコルスタックを指定する。
         *
         * <p>未指定の場合デフォルトのプロトコルスタックが使用される。
         *
         * @param protocol 使用するプロトコルスタック
         * @return このインスタンス
         * @see Protocol
         */
        @Nonnull
        public Params setProtocol(@Nonnull final Protocol protocol) {
            mProtocol = protocol;
            return this;
        }

        @Nonnull
        Protocol getProtocol() {
            return mProtocol;
        }

        /**
         * 使用するインターフェースを指定する。
         *
         * <p>未指定の場合、プロトコルスタックから自動選択される。
         *
         * @param interfaces 使用するインターフェース
         * @return このインスタンス
         */
        @Nonnull
        public Params setInterfaces(@Nullable final Collection<NetworkInterface> interfaces) {
            mInterfaces = interfaces;
            return this;
        }

        /**
         * 使用するインターフェースを指定する。
         *
         * <p>未指定の場合、プロトコルスタックから自動選択される。
         *
         * @param nif 使用するインターフェース
         * @return このインスタンス
         */
        @Nonnull
        public Params setInterface(@Nullable final NetworkInterface nif) {
            mInterfaces = Collections.singletonList(nif);
            return this;
        }

        @Nullable
        Collection<NetworkInterface> getInterfaces() {
            return mInterfaces;
        }

        /**
         * SSDP Notifyパケットを受け取った時にセグメントチェックを行う設定を行う
         *
         * @param enabled セグメントチェックを有効にするときtrue
         * @return このインスタンス
         */
        @Nonnull
        public Params setNotifySegmentCheckEnabled(final boolean enabled) {
            mNotifySegmentCheckEnabled = enabled;
            return this;
        }

        boolean isNotifySegmentCheckEnabled() {
            return mNotifySegmentCheckEnabled;
        }

        /**
         * コールバックを実行する{@link TaskExecutor}を指定する。
         *
         * コールバックのスレッドを指定したい場合に指定する。
         * 未指定の場合singleThreadのExecutorが使用される。
         *
         * @param executor コールバックを実行するTaskExecutor
         * @return このインスタンス
         */
        @Nonnull
        public Params setCallbackExecutor(@Nullable final TaskExecutor executor) {
            mCallbackExecutor = executor;
            return this;
        }

        /**
         * コールバックを実行する{@link CallbackHandler}を指定する。
         *
         * コールバックのスレッドを指定したい場合に指定する。
         * 未指定の場合singleThreadのExecutorが使用される。
         *
         * {@link #setCallbackExecutor(TaskExecutor)}と排他利用であり、後で設定したものが優先される。
         * {@link ControlPoint#terminate()}を実行しても何ら終了処理は行われないため、必要であれば利用側で後始末を実行する。
         *
         * Androidプラットホームであれば、android.os.Handlerを使うことで簡単にUIスレッドでコールバックが実行できる。
         * <pre>
         * param.setCallbackHandler(handler::post)
         * </pre>
         *
         * @param handler コールバックを実行するCallbackHandler
         * @return このインスタンス
         * @see #setCallbackExecutor(TaskExecutor)
         */
        @Nonnull
        public Params setCallbackHandler(@Nonnull final CallbackHandler handler) {
            mCallbackExecutor = new TaskExecutorWrapper(handler);
            return this;
        }

        @Nullable
        TaskExecutor getCallbackExecutor() {
            return mCallbackExecutor;
        }
    }

    private static class TaskExecutorWrapper implements TaskExecutor {
        private final CallbackHandler mHandler;

        TaskExecutorWrapper(@Nonnull final CallbackHandler handler) {
            mHandler = handler;
        }

        @Override
        public boolean execute(@Nonnull final Runnable task) {
            return mHandler.execute(task);
        }

        @Override
        public void terminate() {
        }
    }

    /**
     * ControlPointのインスタンスを作成する。
     *
     * <p>引数のインターフェースを利用するように初期化される。
     * 使用するインターフェースは自動的に選定される。
     *
     * @return ControlPointのインスタンス
     * @throws IllegalStateException 使用可能なインターフェースがない。
     */
    @Nonnull
    public static ControlPoint create()
            throws IllegalStateException {
        return create(new Params());
    }

    /**
     * ControlPointのインスタンスを作成する。
     *
     * @param params 初期化パラメータ
     * @return ControlPointのインスタンス
     * @throws IllegalStateException 使用可能なインターフェースがない。
     */
    @Nonnull
    public static ControlPoint create(@Nonnull final Params params) {
        final Protocol protocol = params.getProtocol();
        return new ControlPointImpl(
                protocol,
                getDefaultInterfacesIfEmpty(protocol, params.getInterfaces()),
                params.isNotifySegmentCheckEnabled(),
                new DiFactory(protocol, params.getCallbackExecutor()));
    }

    @Nonnull
    private static Collection<NetworkInterface> getDefaultInterfacesIfEmpty(
            @Nonnull final Protocol protocol,
            @Nullable final Collection<NetworkInterface> interfaces) {
        if (interfaces == null || interfaces.isEmpty()) {
            return protocol.getAvailableInterfaces();
        }
        return interfaces;
    }

    // インスタンス化禁止
    private ControlPointFactory() {
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/ControlPointFactory.kt
/*
 * Copyright (c) 2018 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp

import net.mm2d.upnp.Adapter.taskExecutor
import net.mm2d.upnp.internal.impl.ControlPointImpl
import net.mm2d.upnp.internal.impl.DiFactory
import java.net.NetworkInterface

/**
 * Factory to create an instance of ControlPoint.
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
object ControlPointFactory {
    /**
     * Create an instance of ControlPoint.
     *
     * @param protocol protocol stack.
     * If not specified, the default protocol stack is used.
     * @param interfaces network interface.
     * If not specified, it is automatically selected from the protocol stack.
     * @param callbackExecutor callback executor.
     * All callbacks are executed from the callback thread.
     * This argument is used to specify the callback thread.
     * If not specified, single Thread Executor is used.
     * @param callbackHandler callback handler.
     * This is used as a TaskExecutor.
     * All callbacks are executed from the callback thread.
     * This argument is used to specify the callback thread.
     * If not specified, single Thread Executor is used.
     * @param notifySegmentCheckEnabled Set whether to check the segment of SSDP Notify packet or not.
     * @param subscriptionEnabled Set whether to use event subscription.
     * Default is true.
     * If false, reduce memory and number of threads at the expense of disabling the feature of event subscription.
     * In that state, when the method for subscription are called, IllegalStateException will be thrown.
     * @param multicastEventingEnabled set whether to use multicast eventing.
     * Default is false.
     * @return Instance of ControlPoint.
     * @throws IllegalStateException There is no interface available.
     */
    @JvmStatic
    fun create(
        protocol: Protocol = Protocol.DEFAULT,
        interfaces: Iterable<NetworkInterface>? = null,
        callbackExecutor: TaskExecutor? = null,
        callbackHandler: ((Runnable) -> Boolean)? = null,
        notifySegmentCheckEnabled: Boolean = false,
        subscriptionEnabled: Boolean = true,
        multicastEventingEnabled: Boolean = false
    ): ControlPoint {
        val executor = callbackExecutor
            ?: callbackHandler?.let { taskExecutor(it) }
        return ControlPointImpl(
            protocol,
            getDefaultInterfacesIfEmpty(protocol, interfaces),
            notifySegmentCheckEnabled,
            subscriptionEnabled,
            multicastEventingEnabled,
            DiFactory(protocol, executor)
        )
    }

    private fun getDefaultInterfacesIfEmpty(
        protocol: Protocol,
        interfaces: Iterable<NetworkInterface>?
    ): Iterable<NetworkInterface> {
        return if (interfaces?.none() != false) {
            protocol.getAvailableInterfaces()
        } else interfaces
    }

    /**
     * Return Builder instance.
     */
    @JvmStatic
    fun builder(): ControlPointBuilder = ControlPointBuilder()

    /**
     * Builder for ControlPoint.
     *
     * Mainly assuming to be use from Java.
     */
    class ControlPointBuilder internal constructor() {
        private var protocol: Protocol = Protocol.DEFAULT
        private var interfaces: Iterable<NetworkInterface>? = null
        private var callbackExecutor: TaskExecutor? = null
        private var notifySegmentCheckEnabled: Boolean = false
        private var subscriptionEnabled: Boolean = true
        private var multicastEventingEnabled: Boolean = false

        /**
         * Set protocol stack.
         *
         * If not specified, the default protocol stack is used.
         *
         * @param proto Protocol stack
         * @return builder
         */
        fun setProtocol(proto: Protocol): ControlPointBuilder = apply {
            protocol = proto
        }

        /**
         * Set network interface.
         *
         * If not specified, it is automatically selected from the protocol stack.
         *
         * @param ifs Network interface
         * @return builder
         */
        fun setInterfaces(ifs: Iterable<NetworkInterface>?): ControlPointBuilder = apply {
            interfaces = ifs
        }

        /**
         * Set callback executor.
         *
         * All callbacks are executed from the callback thread.
         * This argument is used to specify the callback thread.
         * If not specified, single Thread Executor is used.
         *
         * @param executor callback executor
         * @return builder
         */
        fun setCallbackExecutor(executor: TaskExecutor): ControlPointBuilder = apply {
            callbackExecutor = executor
        }

        /**
         * Set callback handler.
         *
         * This is used as a TaskExecutor.
         * All callbacks are executed from the callback thread.
         * This argument is used to specify the callback thread.
         * If not specified, single Thread Executor is used.
         *
         * @param handler callback executor
         * @return builder
         */
        fun setCallbackHandler(handler: (Runnable) -> Boolean): ControlPointBuilder = apply {
            callbackExecutor = taskExecutor(handler)
        }

        /**
         * Set whether to check the segment of SSDP Notify packet or not.
         *
         * @param enabled true, check the segment. false, otherwise
         * @return builder
         */
        fun setNotifySegmentCheckEnabled(enabled: Boolean): ControlPointBuilder = apply {
            notifySegmentCheckEnabled = enabled
        }

        /**
         * Set whether to use event subscription.
         *
         * Default is true.
         * If set to false, reduce memory and number of threads at the expense of disabling the feature of event subscription.
         * In that state, when the method for subscription are called, IllegalStateException will be thrown.
         *
         * @param enabled true, enable event subscription. false, otherwise
         * @return builder
         */
        fun setSubscriptionEnabled(enabled: Boolean): ControlPointBuilder = apply {
            subscriptionEnabled = enabled
        }

        /**
         * Set whether to use multicast eventing.
         *
         * **Multicast eventing is an experimental feature.**
         * Because no other implementation has been found and compatibility has not been verified at all.
         *
         * Default is false.
         * If set to true, multicast events can be received.
         *
         * @param enabled true, enable multicast eventing. false, otherwise
         * @return builder
         */
        fun setMulticastEventingEnabled(enabled: Boolean): ControlPointBuilder = apply {
            multicastEventingEnabled = enabled
        }

        /**
         * Build an instance of ControlPoint.
         *
         * @return instance of ControlPoint
         * @throws IllegalStateException There is no interface available.
         */
        fun build(): ControlPoint = ControlPointImpl(
            protocol,
            getDefaultInterfacesIfEmpty(protocol, interfaces),
            notifySegmentCheckEnabled,
            subscriptionEnabled,
            multicastEventingEnabled,
            DiFactory(protocol, callbackExecutor)
        )
    }
}
